airflow/triggers/temporal.py

```diff
@@ -34,18 +35,30 @@ class DateTimeTrigger(BaseTrigger):
     The provided datetime MUST be in UTC.
     """
 
-    def __init__(self, moment: datetime.datetime):
+    def __init__(self, moment: datetime.datetime, *, soft_fail: bool = False) -> None:
         super().__init__()
+        skipping_message_postfix = "Skipping due to soft_fail is set to True."
         if not isinstance(moment, datetime.datetime):
-            raise TypeError(f"Expected datetime.datetime type for moment. Got {type(moment)}")
+            message = f"Expected datetime.datetime type for moment. Got {type(moment)}"
+            if soft_fail:
+                raise AirflowSkipException(f"{message}. {skipping_message_postfix}")
+            raise TypeError(message)
         # Make sure it's in UTC
         elif moment.tzinfo is None:
-            raise ValueError("You cannot pass naive datetimes")
+            message = "You cannot pass naive datetimes"
+            if soft_fail:
+                raise AirflowSkipException(f"{message}. {skipping_message_postfix}")
```

Comment:
How could raising an `AirflowSkipException` in the trigger pass the task state to skipped? The trigger exception is not propagated to task resuming if I am not mistaken. WDYT?

Reply:
Take `DateTimeSensorAsync` as an example. The second line will raise the `AirflowSkipException` before we execute `self.defer`

```python
        self.defer(
            trigger=DateTimeTrigger(moment=target_dttm, soft_fail=self.soft_fail),
            method_name="execute_complete",
        )
```

Reply:
What about:
```python
         try:
           self.defer(
              trigger=DateTimeTrigger(moment=target_dttm, soft_fail=self.soft_fail),
              method_name="execute_complete",
          )
        except Exception as e:
          raise AirflowException(str(e))
```
or just changing `raise TypeError(message)` and `raise ValueError` to `AirflowException`.

In both cases the exception will be cached by BaseSensor, and a `AirflowSkipException` will be raised when soft_fail is True. I'm trying to avoid complicating the trigger code, and duplicating the soft fail logic.

cc: @uranusjr @potiuk WDYT?

Reply:
Or what do you think about my first commit? https://github.com/apache/airflow/pull/33403/commits/b2f2662ae1a11ea928aad57acd2892c763c2db25 I was trying not to duplicate the check logic and thus moved it to the trigger

Reply:
for me it's much better